app/services/DFS.coffee

```diff
@@ -0,0 +1,72 @@
+AdjacentCellsCalculator = require "./AdjacentCellsCalculator"
+GameGrid                = require "../models/GameGrid"
+LastInColumn            = require "./LastInColumn"
+Tuple                   = require "../models/Tuple"
+TupleSet                = require "../models/TupleSet"
+
+class DFS
+
+  @shuffle: (array) ->
+    # Fisher-Yates shuffle
+    m = array.length
+    t = undefined
+    i = undefined
+    # While there remain elements to shuffle…
+    while m
+      # Pick a remaining element…
+      i = Math.floor(Math.random() * m--)
+      # And swap it with the current element.
+      t = array[m]
+      array[m] = array[i]
+      array[i] = t
+    array
+
+  @search: (seed, input, takenCells) ->
+    return true if input.length is 0
+
+    toVisit = (new AdjacentCellsCalculator(@grid, null, seed.x, seed.y))
+    toVisit = @shuffle toVisit.calculate(takenCells.list)
+    return false if toVisit.length is 0
+    curr = toVisit.shift()
+    return false if (new LastInColumn).isLastAndBlocking @grid.grid, curr.x, curr.y
```

Comment:
Could inject this dependency, so the constructor would be
```
@search: (seed, input, takenCells, last={new LastInColumn}) ->
```
then line 31 would be
```
return false if last.isLastAndBlocking ....
```

Sanity-check my syntax and variable names